### ISSUE TITLE
AWS: EBS volume mount failures due to "timeout expired waiting for volumes to attach/mount"

### ISSUE THREAD

**Issue by argusua** [kubernetes/kubernetes#28892]
EBS Volume can't be attached after minion replacement.
Kubernetes v1.3.0

**Test Case**
1. Create EBS volume with aws ec2 create-volume
2. Launch a replication controller with replicas=1 that references the volume.
3. Discover which ec2 instance (minion) the pod is running on.
4. Mark this instance as unhealthy in autoscaling group and wait for startup new instance.

**Expected outcome**
- The new pod successfully starts on new instance
- The volume is attached successfully.

**Actual outcome**

-  The new pod is scheduled to a new instance
- The pod is stuck in "Pending" because there was a transient failure attaching the EBS volume.

Output from describe pod:
```
Name:           mongo-rs1-ub9yb
Namespace:      test
Node:           ip-172-20-2-143.eu-west-1.compute.internal/172.20.2.143
Start Time:     Wed, 13 Jul 2016 10:23:24 +0300
Labels:         app=mongodb
                environment=test
                role=mongo-rs1
Status:         Pending
IP:
Controllers:    ReplicationController/mongo-rs1
Containers:
  mongo-rs1:
    Container ID:
    Image:              mongo
    Image ID:
    Port:               27017/TCP
    Command:
      mongod
      --replSet
      oprint
      --smallfiles
      --noprealloc
    State:                      Waiting
      Reason:                   ContainerCreating
    Ready:                      False
    Restart Count:              0
    Environment Variables:      <none>
Conditions:
  Type          Status
  Initialized   True 
  Ready         False 
  PodScheduled  True 
Volumes:
  mongo-persistent-storage:
    Type:       AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:   vol-84d7b507
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
  default-token-w7kvi:
    Type:       Secret (a volume populated by a Secret)
    SecretName: default-token-w7kvi
QoS Tier:       BestEffort
Events:
  FirstSeen     LastSeen        Count   From                                                    SubobjectPath   Type            Reason          Message
  ---------     --------        -----   ----                                                    -------------   --------        ------          -------
  4h            1m              115     {kubelet ip-172-20-2-143.eu-west-1.compute.internal}                    Warning         FailedMount     Unable to mount volumes for pod "mongo-rs1-ub9yb_test(aeb3cc37-48ca-11e6-ad15-0a23686b1931)": timeout expired waiting for volumes to attach/mount for pod "mongo-rs1-ub9yb"/"test". list of unattached/unmounted volumes=[mongo-persistent-storage]
  4h            1m              115     {kubelet ip-172-20-2-143.eu-west-1.compute.internal}                    Warning         FailedSync      Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "mongo-rs1-ub9yb"/"test". list of unattached/unmounted volumes=[mongo-persistent-storage]
```

Output from /var/log/kube-controller-manager.log

```
I0713 11:49:29.651482       5 event.go:216] Event(api.ObjectReference{Kind:"ReplicationController", Namespace:"test", Name:"mongo-rs1", UID:"68d8149b-4831-11e6-ad15-0a23686b1931", APIVersion:"v1", ResourceVersion:"52585", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: mongo-rs1-bjq3h
W0713 11:50:30.429047       5 request.go:347] Field selector: v1 - serviceaccounts - metadata.name - default: need to check if this is versioned correctly.
I0713 11:50:34.769939       5 reconciler.go:134] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-c9b14f3c" from node "ip-172-20-1-21.eu-west-1.compute.internal" due to maxWaitForUnmountDuration expiry.
I0713 11:50:34.770035       5 reconciler.go:134] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-84d7b507" from node "ip-172-20-2-7.eu-west-1.compute.internal" due to maxWaitForUnmountDuration expiry.
E0713 11:50:34.887545       5 attacher.go:190] Error checking if volume ("vol-c9b14f3c") is already attached to current node ("ip-172-20-1-21.eu-west-1.compute.internal"). Will continue and try detach anyway. err=error finding instance ip-172-20-1-21.eu-west-1.compute.internal: no instances found for name: ip-172-20-1-21.eu-west-1.compute.internal
E0713 11:50:34.920696       5 attacher.go:190] Error checking if volume ("vol-84d7b507") is already attached to current node ("ip-172-20-2-7.eu-west-1.compute.internal"). Will continue and try detach anyway. err=error finding instance ip-172-20-2-7.eu-west-1.compute.internal: no instances found for name: ip-172-20-2-7.eu-west-1.compute.internal
E0713 11:50:34.949350       5 attacher.go:200] Error detaching volumeID "vol-c9b14f3c": error finding instance ip-172-20-1-21.eu-west-1.compute.internal: no instances found for name: ip-172-20-1-21.eu-west-1.compute.internal
E0713 11:50:34.949405       5 goroutinemap.go:155] Operation for "kubernetes.io/aws-ebs/vol-c9b14f3c" failed. No retries permitted until 2016-07-13 11:52:34.949398367 +0000 UTC (durationBeforeRetry 2m0s). error: DetachVolume.Detach failed for volume "kubernetes.io/aws-ebs/vol-c9b14f3c" (spec.Name: "mongo-persistent-storage") from node "ip-172-20-1-21.eu-west-1.compute.internal" with: error finding instance ip-172-20-1-21.eu-west-1.compute.internal: no instances found for name: ip-172-20-1-21.eu-west-1.compute.internal
E0713 11:50:34.982511       5 attacher.go:200] Error detaching volumeID "vol-84d7b507": error finding instance ip-172-20-2-7.eu-west-1.compute.internal: no instances found for name: ip-172-20-2-7.eu-west-1.compute.internal
E0713 11:50:34.982555       5 goroutinemap.go:155] Operation for "kubernetes.io/aws-ebs/vol-84d7b507" failed. No retries permitted until 2016-07-13 11:52:34.982549022 +0000 UTC (durationBeforeRetry 2m0s). error: DetachVolume.Detach failed for volume "kubernetes.io/aws-ebs/vol-84d7b507" (spec.Name: "mongo-persistent-storage") from node "ip-172-20-2-7.eu-west-1.compute.internal" with: error finding instance ip-172-20-2-7.eu-west-1.compute.internal: no instances found for name: ip-172-20-2-7.eu-west-1.compute.internal
I0713 11:50:35.571842       5 reconciler.go:134] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-05419cb4" from node "ip-172-20-0-48.eu-west-1.compute.internal" due to maxWaitForUnmountDuration expiry.
E0713 11:50:35.631140       5 attacher.go:190] Error checking if volume ("vol-05419cb4") is already attached to current node ("ip-172-20-0-48.eu-west-1.compute.internal"). Will continue and try detach anyway. err=error finding instance ip-172-20-0-48.eu-west-1.compute.internal: no instances found for name: ip-172-20-0-48.eu-west-1.compute.internal
E0713 11:50:35.705306       5 attacher.go:200] Error detaching volumeID "vol-05419cb4": error finding instance ip-172-20-0-48.eu-west-1.compute.internal: no instances found for name: ip-172-20-0-48.eu-west-1.compute.internal
E0713 11:50:35.705354       5 goroutinemap.go:155] Operation for "kubernetes.io/aws-ebs/vol-05419cb4" failed. No retries permitted until 2016-07-13 11:52:35.705348233 +0000 UTC (durationBeforeRetry 2m0s). error: DetachVolume.Detach failed for volume "kubernetes.io/aws-ebs/vol-05419cb4" (spec.Name: "mongo-persistent-storage") from node "ip-172-20-0-48.eu-west-1.compute.internal" with: error finding instance ip-172-20-0-48.eu-west-1.compute.internal: no instances found for name: ip-172-20-0-48.eu-west-1.compute.internal
```

**Comment by mikelorant** [comment]
Can verify this issue and I am having the exact same problem. Destroying minions with EBS volumes attached does not allow the container to restart.

**Comment by apenney** [comment]
Just ran into the same problem on 1.3.2, except I haven't destroyed any minions.  I just suddenly lost the ability to mount volumes.

**Comment by oxygen0211** [comment]
Exact same problem here. Endet up creating a new EBS volume and migrating the data. 

From the source code side, it looked to me like the volumemanager's cached actual_state_of_world didn't match the real state of worl and so it needs to be recreated. I am, however not a maintainer, so this theory is based on some minutes of looking through the code generating the ogs.

**Comment by argusua** [comment]
Looks like this fixed in 1.3.4 [#29485](https://github.com/kubernetes/kubernetes/pull/29485)
Needs to be checked.

**Comment by argusua** [comment]
Fixed for me in  v1.3.4
